app/src/main/java/com/simply/birthdayapp/presentation/viewmodels/ShopsViewModel.kt

```diff
@@ -0,0 +1,80 @@
+package com.simply.birthdayapp.presentation.viewmodels
+
+import android.util.Log
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.simply.birthdayapp.data.repositories.ShopsRepository
+import com.simply.birthdayapp.presentation.ui.models.Shop
+import kotlin.coroutines.CoroutineContext
+import kotlinx.coroutines.CoroutineExceptionHandler
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.FlowPreview
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.debounce
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+
+@OptIn(FlowPreview::class)
+class ShopsViewModel(
+    private val shopsRepository: ShopsRepository,
+) : ViewModel() {
+    private val ioCatchingCoroutineContext: CoroutineContext = Dispatchers.IO + CoroutineExceptionHandler { _, cause ->
+        Log.d(this::class.java.simpleName, Log.getStackTraceString(cause))
+    }
+
+    private val _cachedShops: MutableList<Shop> = mutableListOf()
+
+    private val _loading: MutableStateFlow<Boolean> = MutableStateFlow(true)
+    private val _shops: MutableStateFlow<List<Shop>> = MutableStateFlow(emptyList())
+    private val _scrollPosition: MutableStateFlow<Int> = MutableStateFlow(0)
+    private val _searchBarQuery: MutableStateFlow<String> = MutableStateFlow("")
+
+    val loading: StateFlow<Boolean> = _loading.asStateFlow()
+    val shops: StateFlow<List<Shop>> = _shops.asStateFlow()
+    val scrollPosition: StateFlow<Int> = _scrollPosition.asStateFlow()
+    val searchBarQuery: StateFlow<String> = _searchBarQuery.asStateFlow()
+
+    init {
+        observeSearchBarQuery()
+        fetchShops()
+    }
+
+    fun onScrollPositionChange(scrollPosition: Int) {
+        _scrollPosition.update { scrollPosition }
+    }
+
+    fun onSearchBarQueryChange(searchBarQuery: String) {
+        _searchBarQuery.update { searchBarQuery }
+    }
+
+    private fun observeSearchBarQuery() {
+        _searchBarQuery
+            .debounce(300L)
+            .distinctUntilChanged()
+            .onEach { filterShops(it) }
+            .launchIn(viewModelScope)
+    }
+
+    private fun fetchShops() {
+        viewModelScope.launch(ioCatchingCoroutineContext) {
+            _loading.update { true }
+            try {
+                if (_cachedShops.isEmpty()) _cachedShops += shopsRepository.getShops()
+            } finally {
+                filterShops("")
```

Comment:
This will be executed even if an error will occur. But do we need that so? For the error we don't have a new list, so we don't need to filter anything. Also, fetchShops is called in the init block, so I guess we can filter the shops after getting a successful response.

Reply:
[Fixed](https://github.com/SimplyTechnologies/android-internship/pull/14/commits/f886a4155dbd813c22dc7592d3097a7239002c43)